src/board.rs

```diff
@@ -44,25 +441,1390 @@ impl EnPassantTarget {
+                player.pawns &= !(1 << to_move.from);
+                match to_move.promotion {
+                Promotion::None => player.pawns |= 1 << to_move.to,
+                Promotion::Knight => player.knights |= 1 << to_move.to,
+                Promotion::Bishop => player.bishops |= 1 << to_move.to,
+                Promotion::Rook => player.rooks |= 1 << to_move.to,
+                Promotion::Queen => player.queens |= 1 << to_move.to,
+                }
+                self.half_counter = 0;
+            }
+            PieceType::Rook(_) => {
+                player.rooks &= !(1 << to_move.from);
+                player.rooks |= 1 << to_move.to;
+                if to_move.from as i16 - player.king.ilog2() as i16 > 0 {
+                    player.queen_castle = false;
+                } else {
+                    player.king_castle = false;
+                }
+            }
+            PieceType::Knight(_) => {
+                player.knights &= !(1 << to_move.from);
+                player.knights |= 1 << to_move.to;
+            }
+            PieceType::Bishop(_) => {
+                player.bishops &= !(1 << to_move.from);
+                player.bishops |= 1 << to_move.to;
+            }
+            PieceType::Queen(_) => {
+                player.queens &= !(1 << to_move.from);
+                player.queens |= 1 << to_move.to;
+            }
+            PieceType::King(active_player) => {
+                if player.king_castle == true && 1 << to_move.to & 0x0200000000000002u64 != 0 {
+                    player.rooks &= !(0x0100000000000001);
+                    player.rooks |= 1 << (to_move.to + 1);
+                }
+                if player.queen_castle == true && 1 << to_move.to & 0x2000000000000020u64 != 0 {
+                    player.rooks &= !(0x8000000000000080);
+                    player.rooks |= 1 << (to_move.to - 1);
+                }
+                player.king &= !(1 << to_move.from);
+                player.king |= 1 << to_move.to;
+
+                player.queen_castle = false;
+                player.king_castle = false;
+                if player.king.count_ones() != 1 {
+                    println!("active player: {:?}", active_player);
+                    println!("actual active player: {:?}", self.active_player);
+                    println!("kings: {}", player.king.count_ones());
+                    for i in 0..=7u8 {
+                        let row = (player.king >> (56 - (i * 8))) as u8;
+                        println!("{:08b}", row);
+                    }
+                    println!("");
+                    for i in 0..=7u8 {
+                        let row = (old_king >> (56 - (i * 8))) as u8;
+                        println!("{:08b}", row);
+                    }
+                    //println!("move string: {}", move_str);
+                    println!("move struct: {:?}", to_move);
+                    println!("{:?}", self.move_stack);
+                    panic!();
```

Comment:
we might wanna use [`eprintln!`](https://doc.rust-lang.org/book/ch12-06-writing-to-stderr-instead-of-stdout.html)/`stderr` for error printing rather than the standard `println!`/`stdout`